src/router.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use ceresdbproto::storage::RouteRequest;
+use dashmap::DashMap;
+
+use crate::{
+    errors::Result,
+    model::route::Endpoint,
+    rpc_client::{RpcClient, RpcContext},
+    Error,
+};
+
+/// Used to route metrics to endpoints.
+#[async_trait]
+pub trait Router: Send + Sync {
+    async fn route(&self, metrics: &[String], ctx: &RpcContext) -> Result<Vec<Option<Endpoint>>>;
+
+    fn evict(&self, metrics: &[String]);
+}
+
+/// Implementation for [`Router`].
+///
+/// There is cache in [`RouterImpl`], it will return endpoints in cache first.
+/// If returned endpoints is outdated, you should call [`evict`] to remove them.
+/// And [`RouterImpl`] will fetch new endpoints when you call ['route'] again.
+///
+/// [`route`]: RouterImpl::route
+/// [`evict`]: RouterImpl::evict
+pub struct RouterImpl<R: RpcClient> {
+    cache: DashMap<String, Endpoint>,
+    rpc_client: R,
+}
+
+impl<R: RpcClient> RouterImpl<R> {
+    pub fn new(rpc_client: R) -> Self {
+        Self {
+            cache: DashMap::new(),
+            rpc_client,
+        }
+    }
+}
+
+#[async_trait]
+impl<R: RpcClient> Router for RouterImpl<R> {
+    async fn route(&self, metrics: &[String], ctx: &RpcContext) -> Result<Vec<Option<Endpoint>>> {
+        let mut target_endpoints = vec![None; metrics.len()];
+
+        // Find from cache firstly (if force_refresh is false),
+        // and collect misses.
+        let misses = {
+            let mut misses = HashMap::new();
+            for (idx, metric) in metrics.iter().enumerate() {
+                match self.cache.get(metric) {
+                    Some(pair) => {
+                        target_endpoints[idx] = Some(pair.value().clone());
+                    }
+
+                    None => {
+                        // There should not be duplicated metric in metrics
+                        if misses.insert(metric.clone(), idx).is_some() {
+                            return Err(Error::Unknown(format!(
+                                "Route duplicated metric:{}",
+                                metric
+                            )));
```

Comment:
Is it necessary to check this?

Reply:
Normally, metric in metrics should be unique and it is ensured before they are passed to this function. 
Maybe use assert to replace such error?

Reply:
What will happen if metrics have duplicates and we don't check them? At least in this function everything will be ok I think.